### DISCUSSION TITLE
Is it possible to build without RTTI Support?

### DISCUSSION THREAD

**Discussion by jwillikers** [boostorg/geometry#1041] [Q&A]
Is it possible to use Boost.Geometry without requiring RTTI support? There's a `dynamic_cast` [here](https://github.com/boostorg/geometry/blob/070e3517c4ba3e1c897e962a82072cd4b59e7c19/include/boost/geometry/index/detail/rtree/query_iterators.hpp#L250) that makes it impossible to build with RTTI disabled. Thanks!

```
include/boost/geometry/index/detail/rtree/query_iterators.hpp:250:44: error: ‘dynamic_cast’ not permitted with ‘-fno-rtti’
  250 |         const query_iterator_wrapper * p = dynamic_cast<const query_iterator_wrapper *>(boost::addressof(r));
```

**Comment by mloskot** [comment]
The library is header-only, so there is no building involved.

You can pick the library features selectively via including relevant headers - it is not mandatory to include the all-in-one `<boost/geometry.hpp>`.

So, if you don't use the spatial index features, i.e. you don't include any `<boost/geometry/index/*>` then you should be able to build w/o RTTI enabled. Have you tried that?

**Reply by jwillikers** [reply]
Unfortunately, we use `rtree_type::const_query_iterator` from `boost/geometry/index/rtree.hpp`.

**Reply by awulkiew** [reply]
Would you like to disable it in debug builds too?

**Reply by jwillikers** [reply]
Yes, it's an embedded application with pretty tight memory requirements so we disable RTTI for both debug and release builds.

**Reply by awulkiew** [reply]
This is a mechanism preventing comparison of different kind of iterators because these iterators are type-erased, because with pre-c++20 compilers (without sentinels) begin and end has to have the same type and to not force the user to pass predicates to `qend()` function. This way it's possible to create a query range and pass it to c++11 range-based for loop, etc. 

I guess if the function in question is not instantiated then the error is not generated?
There are versions of `qbegin_` and `qend_` which don't return type-erased iterators:
https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/index/rtree.hpp#L1244
https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/index/rtree.hpp#L1347
You have to `#define BOOST_GEOMETRY_INDEX_DETAIL_EXPERIMENTAL` to use them. Would that be an acceptable solution?
Just use `auto` as a return type and have in mind that `end_()` returns a sentinel of different type than iterator returned by `qbegin_()`.
This could also be faster since no type erasure mechanism (here virtual functions) would be used.

The only other thing I can think of is to add an `#ifdef` around `dynamic_cast` replacing it with `static_cast` when something like `BOOST_GEOMETRY_DISABLE_RTTI` is defined.

**Reply by jwillikers** [reply]
Defining `BOOST_GEOMETRY_INDEX_DETAIL_EXPERIMENTAL` sounds like a good solution.
I've include Boost Serialization, since that appears to be required, and made it to the following error:

```
example.cpp: In member function ‘auto ExampleClass::example_function(const QRectF&) const’:
example.cpp:65:55: error: ‘boost::geometry::index::rtree<Value, Options, IndexableGetter, EqualTo, Allocator>::query_iterator_t<Predicates> boost::geometry::index::rtree<Value, Options, IndexableGetter, EqualTo, Allocator>::qbegin_(const Predicates&) const [with Predicates = boost::geometry::index::detail::predicates::spatial_predicate<QRectF, boost::geometry::index::detail::predicates::intersects_tag, false>; Value = std::pair<QRectF, Canvas::Item*>; Parameters = boost::geometry::index::linear<10>; IndexableGetter = boost::geometry::index::indexable<std::pair<QRectF, Canvas::Item*> >; EqualTo = boost::geometry::index::equal_to<std::pair<QRectF, Canvas::Item*> >; Allocator = boost::container::new_allocator<std::pair<QRectF, Canvas::Item*> >; boost::geometry::index::rtree<Value, Options, IndexableGetter, EqualTo, Allocator>::query_iterator_t<Predicates> = boost::geometry::index::detail::rtree::iterators::spatial_query_iterator<boost::geometry::index::rtree<std::pair<QRectF, Canvas::Item*>, boost::geometry::index::linear<10> >::members_holder, boost::geometry::index::detail::predicates::spatial_predicate<QRectF, boost::geometry::index::detail::predicates::intersects_tag, false> >]’ is private within this context
         return m_tree.qbegin_( bgi::intersects( box ) );
```

**Reply by awulkiew** [reply]
It seems I broke it in this commit: https://github.com/boostorg/geometry/commit/f838b88a9d852aba997fc073df2ca4ffaf7ebdd0 by adding `private:` here: https://github.com/boostorg/geometry/blob/070e3517c4ba3e1c897e962a82072cd4b59e7c19/include/boost/geometry/index/rtree.hpp#L1178
so you'd have to modify the file to use it, e.g. replace this: https://github.com/boostorg/geometry/blob/070e3517c4ba3e1c897e962a82072cd4b59e7c19/include/boost/geometry/index/rtree.hpp#L1187-L1189
with:
```
#ifdef BOOST_GEOMETRY_INDEX_DETAIL_EXPERIMENTAL
public:
#else
private:
#endif
```

Furthermore you probably shouldn't have to include serialization to use these query iterators so there should be different defines for different experimental features.

**Reply by awulkiew** [reply]
Fix: https://github.com/boostorg/geometry/pull/1042

**Reply by jwillikers** [reply]
@awulkiew Thanks! Your changes work great and remove the need to include Boost Serialization!
